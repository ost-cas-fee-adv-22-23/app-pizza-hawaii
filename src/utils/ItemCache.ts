type TItemCacheCache<T> = Record<string, TItemCacheEntry<T>>;

type TItemCacheEntry<T> = {
	createdAt: number;
	data: T;
};

export class ItemCache<T> {
	private readonly cache: TItemCacheCache<T> = {};
	private readonly ttl: number;

	constructor(ttl: number) {
		this.ttl = ttl;
	}

	add(item: T & { id: string }): void {
		this.cache[item.id] = {
			createdAt: Date.now(),
			data: item,
		};
	}

	remove(id: string): void {
		delete this.cache[id];
	}

	get(id: string): T | false {
		const cacheEntry = this.cache[id];

		// Check if user is already in cache otherwise return false
		if (!cacheEntry) return false;

		// Check if cache entry is expired - if it is, remove it from the cache and return false
		const isExpired = Date.now() - cacheEntry.createdAt > this.ttl;
		if (isExpired) {
			delete this.cache[id];
			return false;
		}

		// Return cached data
		return cacheEntry.data;
	}
}
