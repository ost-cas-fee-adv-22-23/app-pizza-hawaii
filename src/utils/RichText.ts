// Description: This file contains a function that converts plain text to rich text (HTML) with links, mentions, hashtags, etc.

export type RichtextSettings = {
	basics?: boolean;
	links?: boolean;
	markdownLinks?: boolean;
	tags?: boolean;
	mentions?: boolean;
	tagLinkPattern?: string;
	mentionLinkPattern?: string;
};

// Default settings for the Richtext parser (we export them so that you can use them as a base for your own settings)
export const defaultSettings: RichtextSettings = {
	basics: true,
	links: true,
	markdownLinks: true,
	tags: true,
	mentions: true,
	tagLinkPattern: '/tag/$1',
	mentionLinkPattern: '/user/$3',
};

export const parse = (plainText: string, settings: RichtextSettings = defaultSettings): string => {
	let richText = plainText;

	if (settings.markdownLinks) {
		// Convert Markdown-Links in the text and replace with anchor tags (e.g. [Link](https://example.com)) to <a href="https://example.com">Link</a>
		richText = richText.replace(/\[(.*?)\]\((.*?)\)/g, '<a href="$2">$1</a>');
	}

	if (settings.links) {
		// Convert plain text URLs to anchor tags (e.g. https://example.com) to <a href="https://example.com">https://example.com</a> do not change already existing links
		richText = richText.replace(/(?<!href=")(https?:\/\/[^\s]+)/g, '<a href="$1">$1</a>');
	}

	if (settings.tags) {
		// Convert Hashtag-Links in the text and wrap with anchor tags (e.g. #tag) to <a href="/tag/tag">#tag</a>
		richText = richText.replace(/#(\w+)/g, `<a href="${settings.tagLinkPattern}">#$1</a>`);
	}

	if (settings.mentions) {
		// Convert @-mentions in the text and wrap with anchor tags. (e.g. @[JohnDoe|123]) to <a href="/user/123">@JohnDoe</a>
		richText = richText.replace(/@(\[(\w+)\|(\d+)\])/g, `<a href="${settings.mentionLinkPattern}">@$2</a>`);
	}

	if (settings.basics) {
		// Add paragraph tags to the text
		richText = `<p>${richText.split('\n\n').join('</p><p>')}</p>`;

		// Replace newline characters with line break tags
		richText = richText.replace(/\n/g, '<br>');
	}

	return richText;
};
