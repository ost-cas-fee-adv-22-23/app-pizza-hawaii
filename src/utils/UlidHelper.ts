import { decodeTime, encodeTime } from 'ulid';

export const getCurrent = () => encodeTime(new Date().getTime(), 10) + '0000000000000000';
export const getNewerThan = (ulid: string) => encodeTime(decodeTime(ulid) + 1, 10) + '0000000000000000';
export const getOlderThan = (ulid: string) => encodeTime(decodeTime(ulid) - 1, 10) + '0000000000000000';

const UlidHelper = {
	getCurrent,
	getNewerThan,
	getOlderThan,
};

export default UlidHelper;
