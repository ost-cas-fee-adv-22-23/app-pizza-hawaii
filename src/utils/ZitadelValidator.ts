import { TZitadelProfile } from '../types/Zitadel';

export type TGenderOption = {
	key: TZitadelProfile['gender'];
	text: string;
};

export const GenderOptions: TGenderOption[] = [
	{
		key: 'GENDER_UNSPECIFIED',
		text: '-',
	},
	{
		key: 'GENDER_FEMALE',
		text: 'Frau',
	},
	{
		key: 'GENDER_MALE',
		text: 'Herr',
	},
	{
		key: 'GENDER_DIVERSE',
		text: 'Divers',
	},
];
