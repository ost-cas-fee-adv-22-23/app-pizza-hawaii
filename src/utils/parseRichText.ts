export type RichtextSettings = {
	basics?: boolean;
	links?: boolean;
	markdownLinks?: boolean;
	tags?: boolean;
	mentions?: boolean;
	tagLinkPattern?: string;
	mentionLinkPattern?: string;
};

const defaultSettings: RichtextSettings = {
	basics: true,
	links: true,
	markdownLinks: true,
	tags: true,
	mentions: false, // We can't use this feature because we don't have a username (yet)
	tagLinkPattern: '/tag/$1',
	mentionLinkPattern: '/user/$1',
};

const parseRichText = (plainText: string, settings: RichtextSettings = defaultSettings): string => {
	let richText = plainText;

	if (settings.links) {
		// Use a regular expression to find URLs in the text and replace URLs with anchor tags
		richText = richText.replace(/((https?:\/\/|www\.)[^\s]+)/g, '<a href="//$1">$1</a>');
	}

	if (settings.markdownLinks) {
		// Use a regular expression to find Markdown-Links in the text and replace with anchor tags
		richText = richText.replace(/\[(.*?)\]\((.*?)\)/g, '<a href="$2">$1</a>');
	}

	if (settings.tags) {
		// Use a regular expression to find Hashtag-Links in the text and wrap with anchor tags
		richText = richText.replace(/#(\w+)/g, `<a href="${settings.tagLinkPattern}">#$1</a>`);
	}

	if (settings.mentions) {
		// Use a regular expression to find @-mentions in the text and wrap with anchor tags
		richText = richText.replace(/@(\w+)/g, `<a href="${settings.mentionLinkPattern}">@$1</a>`);
	}

	if (settings.basics) {
		// Add paragraph tags to the text
		richText = `<p>${richText.split('\n\n').join('</p><p>')}</p>`;

		// Replace newline characters with line break tags
		richText = richText.replace(/\n/g, '<br>');
	}

	return richText;
};

export default parseRichText;
