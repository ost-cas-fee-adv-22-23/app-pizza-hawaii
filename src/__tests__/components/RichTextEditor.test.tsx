import { cleanup, render, screen } from '@testing-library/react';
import { parse as parseRichText } from '../../utils/RichText';
import { Richtext } from '@smartive-education/pizza-hawaii';


export const defaultSettings = {
	basics: true,
	links: true,
	markdownLinks: true,
	tags: true,
	mentions: true,
	tagLinkPattern: '/tag/$1',
	mentionLinkPattern: '/user/$3',
};

const inputDataHashTag = 'hi there, I am a great fan of pizza hawaii. #pizza #hawaii';
const inputDataLink = 'hi there, I want to share that https://example.com is a great website.';
const inputDataUserMention = 'hi there, I want to share that @[Testuser|214652397815857409] is a crappy website.';
const inputDataMarkdown = 'hi there, I want to share that [lookatthat](https://lookatthat.ch) is a simple website.';
const inputDatawithBreaks = 'hi there, I want to share that long text with a line- \n, break and a pizza slize üçï.';

describe('Richtext Component input rendering', () => {
    afterEach(cleanup);
    // test if richtext component renders correctly
    it('should render the Richtext component with text', () => {
        render(<Richtext size="M" as="div">{inputDataHashTag}</Richtext>);
        expect(screen.getByText('hi there, I am a great fan of pizza hawaii. #pizza #hawaii'));
    });

    // test if link is rendered correctly
    it('should render the Richtext component with text and link with `a-tags`', () => { 
        render(<Richtext size="M" as="div">{inputDataLink}</Richtext>);
        expect(screen.findAllByRole('link', { name: 'https://example.com' }));
    });

    // test if the Richtext renders correct html tags if `as` prop is specified
    it('should render the Richtext component as `article` when specified', () => {
        const { container } = render(<Richtext size="M" as="article">{inputDataLink}</Richtext>);
        expect(container.querySelector('article'));
    });

    // same test as above but with `section` instead of `article`
    it('should render the Richtext component as `section` when specified', () => {
        const { container } = render(<Richtext size="M" as="section">{inputDataLink}</Richtext>);
        expect(container.querySelector('section'));
    });

    // test if the Richtext component renders correct font specific font classes with `size` prop is specified
    it('should render the Richtext component with `size` prop specified as `M` to have the correct css class', () => {
        const { container } = render(<Richtext size="M" as="div">{inputDataLink}</Richtext>);
        expect(container.getElementsByClassName('font-medium leading-normal text-l')).toHaveLength(1);
    });

    // test snapshot
    it('should match the snapshot', () => {
        const { container } = render(<Richtext size="M" as="div">{inputDataLink}</Richtext>);
        expect(container).toMatchSnapshot();
    });
});

// lets test the parse function
describe('test parseRichText functions', () => {
    afterEach(cleanup);
    // test plain text URLs to anchor tags creation when links are posted 
    it('should return link anchor tag when links are posted, clutched by `paragraph-tags`', () => {
        const result = parseRichText(inputDataLink);
        expect(result).toBe('<p>hi there, I want to share that <a href="https://example.com">https://example.com</a> is a great website.</p>');
    });

    // test link creation when hash tags are used
    it('should return link creation when hash tags are used, clutched by `paragraph tags`', () => {
        const result = parseRichText(inputDataHashTag);
        expect(result).toBe('<p>hi there, I am a great fan of pizza hawaii. <a href="/tag/pizza">#pizza</a> <a href="/tag/hawaii">#hawaii</a></p>');
    })

    // test user mentions link creation
    it('should return links to the corresponding user for all user mentions, clutched by `paragraph tags`', () => {
        const result = parseRichText(inputDataUserMention);
        expect(result).toBe('<p>hi there, I want to share that <a href=\"/user/214652397815857409\">@Testuser</a> is a crappy website.</p>');
    })

    // test markdown link creation
    it('should return created links for all markdown links clutched by `paragraph tags`', () => {
        const result = parseRichText(inputDataMarkdown);
        expect(result).toBe('<p>hi there, I want to share that <a href="https://lookatthat.ch">lookatthat</a> is a simple website.</p>');
    });

    // test line breaks creation renders correctly
    it('should return replace line breaks with a `<br>`, clutched by `paragraph tags`', () => {
        const result = parseRichText(inputDatawithBreaks);
        expect(result).toBe('<p>hi there, I want to share that long text with a line- <br>, break and a pizza slize üçï.</p>');
    })
});
