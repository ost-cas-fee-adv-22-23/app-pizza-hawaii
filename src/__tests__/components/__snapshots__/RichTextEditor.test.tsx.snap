// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Richtext Component input rendering should match the snapshot 1`] = `
<div>
  <div
    class="block font-medium leading-normal text-l [&>p]:mb-2 [&>p:last-child]:mb-0 [&>p>a]:text-violet-600 [&>p>a:hover]:text-violet-400 [&>p]:break-words"
  >
    <p>
      hi there, I want to share that https://example.com is a great website.
    </p>
  </div>
</div>
`;
